lib/Dialect/TritonGPU/Transforms/Utility.cpp

```diff
@@ -21,7 +21,13 @@
 
 namespace mlir {
 
-using namespace triton;
+// We only "import" the symbols that we need to avoid name conflicts.
+using triton::AxisInfo;
+using triton::DialectInferLayoutInterface;
+using triton::JoinOp;
+using triton::ModuleAxisInfoAnalysis;
+using triton::PointerType;
+using triton::SplitOp;
```

Comment:
What do those names conflict with? I would assume using `using namespace triton;` should be enough?

Reply:
These names do not conflict. It's other stuff in the `triton` namespace that conflict. For example, there exist both `triton::detail` and `triton::gpu::detail` iirc as one example and those conflicts were causing the build to fail.

Reply:
What would it look like to make those explicit instead? (I don't see `detail` being used)

Reply:
Hmm I'm not sure I understand but I'm also not very familiar with C++ so maybe I'm missing something. My understanding of the issue I ran into is the following. The code previously looked something like this:

```
namespace mlir {
  // At this point the `mlir` already contains a symbol named `detail` coming from the MLIR codebase via the headers.

  // The following line brings all symbols from within `mlir::triton` in the current scope. That namespace also contains
  // a symbol named `detail`. That symbol now conflicts with `mlir::detail` that is already in scope. Further down in the
  // code when the compiler tries to prove a type expression in a SFINAE bound, it fails because of this ambiguity.
  using namespace triton;
}
```

In this PR, instead of importing all symbols from the `mlir::triton` namespace, I switched to only importing the ones that are used in this file, thus avoiding the name conflict/ambiguity. I'm actually not sure why this does not raise an error with Clang, but then again I've noticed how C++ compilers tend to not be super consistent with the standards sometimes and I'm not actually sure what the correct behavior would be here. This change makes it work for both compilers.

Reply:
what I'm asking is can we make the conflicting names explicit instead? There would be conflict only when trying to use a symbol that would exist in both namespace. I'm not sure I understand where that happens and why we cannot just make those names explicit.

Reply:
We're not explicitly using these symbols. The compiler is running into them while trying to derive a template bound for a SFINAE expression. I'm not super clear on how those work but specifically, the calls to the `walk` function (e.g., [here](https://github.com/triton-lang/triton/blob/8a75feb62663c28810ca07ba7f31b48d3f77ae76/lib/Dialect/TritonGPU/Transforms/Utility.cpp#L155)) result in the error and after looking into it for a bit, I couldn't figure out if there's a way to pass something explicit that would resolve this. If you have any suggestions that'd be great and also quite educational for myself.

Reply:
what's the error?

Reply:
I took the version in this branch and replaced these `using triton::...` statements with the previous `using namespace triton` statement. This is what it results in:

```
external/triton/lib/Dialect/TritonGPU/Transforms/Utility.cpp(157): error C2672: 'mlir::OpState::walk': no matching overloaded function found
external/llvm-project/mlir/include\mlir/IR/OpDefinition.h(193): note: could be 'enable_if<llvm::function_traits<decay<_Ty>::type,std::is_class<T>::value>::num_args==2,RetT>::type mlir::OpState::walk(FnT &&)'
        with
        [
            T=decay<_Ty>::type
        ]
external/triton/lib/Dialect/TritonGPU/Transforms/Utility.cpp(157): note: 'enable_if<llvm::function_traits<decay<_Ty>::type,std::is_class<T>::value>::num_args==2,RetT>::type mlir::OpState::walk(FnT &&)': could not deduce template argument for 'RetT'
        with
        [
            T=decay<_Ty>::type
        ]
external/llvm-project/mlir/include\mlir/IR/Visitors.h(465): note: 'mlir::detail::walk': no overloaded function could convert all the argument types
external/llvm-project/mlir/include\mlir/IR/Visitors.h(398): note: could be 'mlir::WalkResult mlir::detail::walk(mlir::Operation *,llvm::function_ref<mlir::WalkResult (mlir::Operation *,const mlir::WalkStage &)>)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(465): note: 'mlir::WalkResult mlir::detail::walk(mlir::Operation *,llvm::function_ref<mlir::WalkResult (mlir::Operation *,const mlir::WalkStage &)>)': cannot convert argument 2 from 'FnT' to 'llvm::function_ref<mlir::WalkResult (mlir::Operation *,const mlir::WalkStage &)>'
        with
        [
            FnT=mlir::GraphDumper::dump::<lambda_73866c6e32d9f3789994c27a660811ed>
        ]
external/llvm-project/mlir/include\mlir/IR/Visitors.h(465): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
external/llvm-project/mlir/include\mlir/IR/Visitors.h(389): note: or       'void mlir::detail::walk(mlir::Operation *,llvm::function_ref<void (mlir::Operation *,const mlir::WalkStage &)>)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(465): note: 'void mlir::detail::walk(mlir::Operation *,llvm::function_ref<void (mlir::Operation *,const mlir::WalkStage &)>)': cannot convert argument 2 from 'FnT' to 'llvm::function_ref<void (mlir::Operation *,const mlir::WalkStage &)>'
        with
        [
            FnT=mlir::GraphDumper::dump::<lambda_73866c6e32d9f3789994c27a660811ed>
        ]
external/llvm-project/mlir/include\mlir/IR/Visitors.h(465): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
external/llvm-project/mlir/include\mlir/IR/Visitors.h(453): note: or       'enable_if<!std::is_same<ArgT,mlir::Operation*>::value&&std::is_same<RetT,mlir::WalkResult>::value,RetT>::type mlir::detail::walk(mlir::Operation *,FuncTy &&)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(428): note: or       'enable_if<!std::is_same<ArgT,mlir::Operation*>::value&&std::is_same<RetT,void>::value,RetT>::type mlir::detail::walk(mlir::Operation *,FuncTy &&)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(411): note: or       'enable_if<std::is_same<ArgT,mlir::Operation*>::value,RetT>::type mlir::detail::walk(mlir::Operation *,FuncTy &&)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(372): note: or       'enable_if<!is_one_of<ArgT,mlir::Operation*,mlir::Region*,mlir::Block*>::value&&std::is_same<RetT,mlir::WalkResult>::value,RetT>::type mlir::detail::walk(mlir::Operation *,FuncTy &&)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(335): note: or       'enable_if<!is_one_of<ArgT,mlir::Operation*,mlir::Region*,mlir::Block*>::value&&std::is_same<RetT,void>::value,RetT>::type mlir::detail::walk(mlir::Operation *,FuncTy &&)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(311): note: or       'enable_if<is_one_of<ArgT,mlir::Operation*,mlir::Region*,mlir::Block*>::value,RetT>::type mlir::detail::walk(mlir::Operation *,FuncTy &&)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(256): note: or       'mlir::WalkResult mlir::detail::walk(mlir::Operation *,llvm::function_ref<mlir::WalkResult (mlir::Operation *)>,mlir::WalkOrder)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(228): note: or       'mlir::WalkResult mlir::detail::walk(mlir::Operation *,llvm::function_ref<mlir::WalkResult (mlir::Block *)>,mlir::WalkOrder)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(200): note: or       'mlir::WalkResult mlir::detail::walk(mlir::Operation *,llvm::function_ref<mlir::WalkResult (mlir::Region *)>,mlir::WalkOrder)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(170): note: or       'void mlir::detail::walk(mlir::Operation *,llvm::function_ref<void (mlir::Operation *)>,mlir::WalkOrder)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(153): note: or       'void mlir::detail::walk(mlir::Operation *,llvm::function_ref<void (mlir::Block *)>,mlir::WalkOrder)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(136): note: or       'void mlir::detail::walk(mlir::Operation *,llvm::function_ref<void (mlir::Region *)>,mlir::WalkOrder)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(465): note: while trying to match the argument list '(nullptr, FnT)'
        with
        [
            FnT=mlir::GraphDumper::dump::<lambda_73866c6e32d9f3789994c27a660811ed>
        ]
external/llvm-project/mlir/include\mlir/IR/OpDefinition.h(165): note: or       'enable_if<llvm::function_traits<decay<FnT>::type,std::is_class<T>::value>::num_args==1,RetT>::type mlir::OpState::walk(FnT &&)'
        with
        [
            T=decay<FnT>::type
        ]
external/triton/lib/Dialect/TritonGPU/Transforms/Utility.cpp(157): note: 'enable_if<llvm::function_traits<decay<FnT>::type,std::is_class<T>::value>::num_args==1,RetT>::type mlir::OpState::walk(FnT &&)': could not deduce template argument for 'RetT'
        with
        [
            T=decay<FnT>::type
        ]
external/llvm-project/mlir/include\mlir/IR/Visitors.h(465): note: 'mlir::detail::walk': no overloaded function could convert all the argument types
external/llvm-project/mlir/include\mlir/IR/Visitors.h(398): note: could be 'mlir::WalkResult mlir::detail::walk(mlir::Operation *,llvm::function_ref<mlir::WalkResult (mlir::Operation *,const mlir::WalkStage &)>)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(465): note: 'mlir::WalkResult mlir::detail::walk(mlir::Operation *,llvm::function_ref<mlir::WalkResult (mlir::Operation *,const mlir::WalkStage &)>)': cannot convert argument 2 from 'FnT' to 'llvm::function_ref<mlir::WalkResult (mlir::Operation *,const mlir::WalkStage &)>'
        with
        [
            FnT=mlir::GraphDumper::dump::<lambda_73866c6e32d9f3789994c27a660811ed>
        ]
external/llvm-project/mlir/include\mlir/IR/Visitors.h(465): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
external/llvm-project/mlir/include\mlir/IR/Visitors.h(389): note: or       'void mlir::detail::walk(mlir::Operation *,llvm::function_ref<void (mlir::Operation *,const mlir::WalkStage &)>)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(465): note: 'void mlir::detail::walk(mlir::Operation *,llvm::function_ref<void (mlir::Operation *,const mlir::WalkStage &)>)': cannot convert argument 2 from 'FnT' to 'llvm::function_ref<void (mlir::Operation *,const mlir::WalkStage &)>'
        with
        [
            FnT=mlir::GraphDumper::dump::<lambda_73866c6e32d9f3789994c27a660811ed>
        ]
external/llvm-project/mlir/include\mlir/IR/Visitors.h(465): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
external/llvm-project/mlir/include\mlir/IR/Visitors.h(453): note: or       'enable_if<!std::is_same<ArgT,mlir::Operation*>::value&&std::is_same<RetT,mlir::WalkResult>::value,RetT>::type mlir::detail::walk(mlir::Operation *,FuncTy &&)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(428): note: or       'enable_if<!std::is_same<ArgT,mlir::Operation*>::value&&std::is_same<RetT,void>::value,RetT>::type mlir::detail::walk(mlir::Operation *,FuncTy &&)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(411): note: or       'enable_if<std::is_same<ArgT,mlir::Operation*>::value,RetT>::type mlir::detail::walk(mlir::Operation *,FuncTy &&)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(372): note: or       'enable_if<!is_one_of<ArgT,mlir::Operation*,mlir::Region*,mlir::Block*>::value&&std::is_same<RetT,mlir::WalkResult>::value,RetT>::type mlir::detail::walk(mlir::Operation *,FuncTy &&)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(335): note: or       'enable_if<!is_one_of<ArgT,mlir::Operation*,mlir::Region*,mlir::Block*>::value&&std::is_same<RetT,void>::value,RetT>::type mlir::detail::walk(mlir::Operation *,FuncTy &&)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(311): note: or       'enable_if<is_one_of<ArgT,mlir::Operation*,mlir::Region*,mlir::Block*>::value,RetT>::type mlir::detail::walk(mlir::Operation *,FuncTy &&)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(256): note: or       'mlir::WalkResult mlir::detail::walk(mlir::Operation *,llvm::function_ref<mlir::WalkResult (mlir::Operation *)>,mlir::WalkOrder)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(228): note: or       'mlir::WalkResult mlir::detail::walk(mlir::Operation *,llvm::function_ref<mlir::WalkResult (mlir::Block *)>,mlir::WalkOrder)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(200): note: or       'mlir::WalkResult mlir::detail::walk(mlir::Operation *,llvm::function_ref<mlir::WalkResult (mlir::Region *)>,mlir::WalkOrder)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(170): note: or       'void mlir::detail::walk(mlir::Operation *,llvm::function_ref<void (mlir::Operation *)>,mlir::WalkOrder)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(153): note: or       'void mlir::detail::walk(mlir::Operation *,llvm::function_ref<void (mlir::Block *)>,mlir::WalkOrder)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(136): note: or       'void mlir::detail::walk(mlir::Operation *,llvm::function_ref<void (mlir::Region *)>,mlir::WalkOrder)'
external/llvm-project/mlir/include\mlir/IR/Visitors.h(465): note: while trying to match the argument list '(nullptr, FnT)'
        with
        [
            FnT=mlir::GraphDumper::dump::<lambda_73866c6e32d9f3789994c27a660811ed>
        ]
external/llvm-project/mlir/include\mlir/IR/PatternMatch.h(358): error C2872: 'detail': ambiguous symbol
bazel-out/x64_windows-opt/bin/external/triton/include\triton/Dialect/TritonNvidiaGPU/IR/Ops.h.inc(2475): note: could be 'mlir::detail'
bazel-out/x64_windows-opt/bin/external/triton/include\triton/Dialect/Triton/IR/Ops.h.inc(8546): note: or       'mlir::triton::detail'
external/llvm-project/mlir/include\mlir/IR/PatternMatch.h(358): note: the template instantiation context (the oldest one first) is
external/triton/lib/Dialect/TritonGPU/Transforms/Utility.cpp(866): note: see reference to class template instantiation 'mlir::OpRewritePattern<mlir::scf::ForOp>' being compiled
external/triton/lib/Dialect/TritonGPU/Transforms/Utility.cpp(895): error C2672: 'mlir::Block::walk': no matching overloaded function found
external/llvm-project/mlir/include\mlir/IR/Block.h(342): note: could be 'RetT mlir::Block::walk(mlir::Block::iterator,mlir::Block::iterator,FnT &&)'
external/triton/lib/Dialect/TritonGPU/Transforms/Utility.cpp(895): note: 'RetT mlir::Block::walk(mlir::Block::iterator,mlir::Block::iterator,FnT &&)': expects 3 arguments - 1 provided
external/llvm-project/mlir/include\mlir/IR/Block.h(290): note: or       'RetT mlir::Block::walk(FnT &&)'
external/triton/lib/Dialect/TritonGPU/Transforms/Utility.cpp(895): note: 'RetT mlir::Block::walk(FnT &&)': could not deduce template argument for 'ArgT'
external/llvm-project/mlir/include\mlir/IR/Block.h(288): note: 'detail': ambiguous symbol
```

Reply:
@ThomasRaoux is this something you might be open to merging?

Reply:
I think it's pretty cumbersome to have to use `using triton::xxx` for everything and not something that we can expect developers to do; it would keep breaking since we don't have CI that can reproduce the error. Is there a more generic solution?

Reply:
Aside from all these `using`, the rest of the PR is pretty minor and ok to merge

Reply:
Yeah I'm also not happy with the proliferation of these using statements. I'm not very familiar with C++ and when I looked into this, I didn't find an alternative. In some ways, what's happening with MSVC is not unreasonable as there are indeed name conflicts/ambiguity and it's just being stricter than clang. Another option is to fully qualify the names instead of using these `using` statements but the `using` approach seemed more innocuous in terms of code changes.

Reply:
is there any way to repro without MSVC? I'm willing to take a look but don't have MSVC locally :')

Reply:
I am tkaing a look and noticing we already use fully qualified names in some places despite the `using namespace`, and a lot of these symbols have a single use so `using triton::symbol` is definitely too bulky. Would you mind modifying this PR to remove `using triton::` and use fully qualified names where appropriate?

Reply:
Oh yeah definitely. I was thinking of doing that originally but wasn't sure what you'd prefer.

Reply:
Regarding reproducing without MSVC, I looked a bit into whether there's any clang flag or something we could flip to reproduce the stricter behavior but I couldn't find anything. :/

Reply:
@ptillet I replaced all references with fully qualified names and removed the `using` statements.